### PR TITLE
battery_level vs. usable_battery_level #321

### PR DESCRIPTION
**Pull request by ctraber**
As a first step I added usable_battery_level to positions table and summary page.

Sorry, but I'm not familiar with the Elixir yet and I'm sure there are still things missing (translation, test,?)...

Bildschirmfoto 2020-01-03 um 12 35 51

### CODE REVIEW

**Comment by ctraber** [comment]
Tests passing on my dev env now.

Here I get:
12:49:44.528 [error] Postgrex.Protocol (#PID<0.1069.0>) disconnected: ** (DBConnection.ConnectionError) owner #PID<0.3853.0> exited

File: lib/teslamate_web/templates/car/summary.html.leex [hunk 1]
```diff
@@ -158,7 +158,29 @@
...
+                if @settings.unit_of_length == :mi do
+                  "#{Float.round(@summary.rated_battery_range_km/@summary.battery_level * 100 / 1.60934, 2)} mi"
+                else
+                  "#{Float.round(@summary.rated_battery_range_km/@summary.battery_level * 100,2)} km"
+                end
+ 
+              %> 100%)</td>
+          </tr>
+          <% end %>
+          <%= unless is_nil(@summary.usable_battery_level) do %>
```
Comment (adriankumpf) [inline lib/teslamate_web/templates/car/summary.html.leex]: Does it make sense to hide the 'Usable State of Charge' row if `usable_battery_level` == `battery_level`? What do you think?

**Comment by adriankumpf** [comment]
Looks really good! After rerunning the tests everything passes :)

**Comment by ctraber** [comment]
How can you restart the test without commit? Just in case i happens again.

**Comment by adriankumpf** [comment]
There is a "restart build" button on the Travis page. But I'm not sure if it allows you to do that. Ideally, it shouldn't be required to rerun that often. Some tests are sensitive to timing though and therefore can be flaky.

File: lib/teslamate_web/templates/car/summary.html.leex (hunk 1 above)
Comment (ctraber) [reply lib/teslamate_web/templates/car/summary.html.leex]: Sure makes sense. There is no additional information when both are the same.
I'll also try to make a snowflake when difference is higher than 2% I think thats what the TeslaApp does.

Comment (adriankumpf) [reply lib/teslamate_web/templates/car/summary.html.leex]: > I'll also try to make a snowflake when difference is higher than 2% I think thats what the TeslaApp does.

Yeah, that would be great!

**Comment by adriankumpf** [comment]
Excellent. Will merge soon!

**Comment by ctraber** [comment]
Finally it was cold enough for a snowflake. By using usable_battery_level, the projected range is not changing that much with temperature. I think it gives a better degradation overview.

Bildschirmfoto 2020-01-08 um 11 51 33

Bildschirmfoto 2020-01-08 um 12 02 38

**Comment by adriankumpf** [comment]
It took a little longer, but I could finally take a closer look at the PR. I made a just few adjustments:

* Replaced the existing degradation dashboard with your 'projected range' dashboard
* Moved the "(xxx km 100%)" text into a tooltip
* Added a test for the snowflake icon

Thanks again! :)

**Comment by adriankumpf** [comment]
Resolves #321

**Comment by ctraber** [comment]
Will I see the tooltip in iOS Safari?
Most of the time I view it on my phone.

**Comment by adriankumpf** [comment]
Should work with 3904a18e2406089b50fc2c933cfa9baf5f180704

**Comment by ctraber** [comment]
Thanks a lot!
Just tried. Works perfectly.

**Comment by ctraber** [comment]
I tested your new projected range dashboard.
In the second panel the coalesce for usable_battery_level is missing. I would add it but for some reason I do not get the dashboard, I changed in filesystem, after building a new grafana image. 

Do you have an idea why it is not showing the changed dashboard after docker build and restart of container?

**Comment by adriankumpf** [comment]
You are referring to the following query (line 3), right?

```SQL
SELECT
	$__timeGroup(date, '6h') AS time,
	convert_km(sum([[preferred_range]]_battery_range_km) / sum(usable_battery_level) * 100, '$length_unit') AS "Projected Range (using usable_battery_level) [$length_unit]",
	convert_km(sum([[preferred_range]]_battery_range_km) / sum(battery_level) * 100, '$length_unit') AS "Projected Range (using battery_level) [$length_unit]"
FROM
	positions
WHERE
	$__timeFilter(date) and
	car_id = $car_id
GROUP BY
	1
ORDER BY
	1,2  DESC
```

I don't see why `sum(coalesce(usable_battery_level,battery_level))` should be necessary. Could you explain that?

> Do you have an idea why it is not showing the changed dashboard after docker build and restart of container?

Does it show up if you add the dashboard under a new name (and with a different `uid`)?

**Comment by ctraber** [comment]
The null values mess up the graph at the point where database starts having values for usable_battera_level. Maybe sum() with some null values, not sure. 

see the spike with app. 1700 km.
Bildschirmfoto 2020-01-18 um 14 33 55

I can add new dashboards.
But, I copied yours with "save as" changed it and copied the json to the existing file, changed names... But if I re-build the container its again your previous version..
=> changing in file, build... and still the version before file change.

**Comment by adriankumpf** [comment]
Hmm, seems to work fine for me. Anyway, if `coalesce` helps, let's add it back in.

> But, I copied yours with "save as" changed it and copied the json to the existing file, changed names... But if I re-build the container its again your previous version..
=> changing in file, build... and still the version before file change.

Very strange. Sounds like it takes the previous file when creating the image? Maybe try doing the following:

1. Stop the container
2. delete the grafana volume
3. build the image
4. run the new image

**Comment by ctraber** [comment]
Deleting the volume did the trick.
Thanks!

**Pull request merged**